File: packages/core/src/godot/capabilities.ts
/**
 * Command capabilities advertised by a Godot executable's `--help` output.
 *
 * Presence here means advertised support, not operationally verified
 * support: the two states are kept distinct (see `verifiedCapabilities` on
 * the engine profile).
 */
export interface GodotCommandCapabilities {
  readonly editor: boolean;
  readonly projectManager: boolean;
  readonly recoveryMode: boolean;
  readonly headless: boolean;
  readonly projectPath: boolean;
  readonly scene: boolean;
  readonly script: boolean;
  readonly checkOnly: boolean;
  readonly import: boolean;
  readonly quit: boolean;
  readonly quitAfter: boolean;
  readonly lsp: boolean;
  readonly dap: boolean;
  readonly debugServer: boolean;
  readonly buildSolutions: boolean;
  readonly extensionApiDump: boolean;
  readonly extensionApiWithDocsDump: boolean;
  readonly extensionApiValidation: boolean;
  readonly docTool: boolean;
  readonly movieWriting: boolean;
}

export function createEmptyGodotCommandCapabilities(): GodotCommandCapabilities {
  return {
    editor: false,
    projectManager: false,
    recoveryMode: false,
    headless: false,
    projectPath: false,
    scene: false,
    script: false,
    checkOnly: false,
    import: false,
    quit: false,
    quitAfter: false,
    lsp: false,
    dap: false,
    debugServer: false,
    buildSolutions: false,
    extensionApiDump: false,
    extensionApiWithDocsDump: false,
    extensionApiValidation: false,
    docTool: false,
    movieWriting: false,
  };
}

/** Immutable, bounded option set recognized by the help capability parser. */
export const GODOT_KNOWN_OPTIONS: readonly {
  readonly option: string;
  readonly capability: keyof GodotCommandCapabilities;
}[] = [
  { option: "--editor", capability: "editor" },
  { option: "--project-manager", capability: "projectManager" },
  { option: "--recovery-mode", capability: "recoveryMode" },
  { option: "--headless", capability: "headless" },
  { option: "--path", capability: "projectPath" },
  { option: "--scene", capability: "scene" },
  { option: "--script", capability: "script" },
  { option: "--check-only", capability: "checkOnly" },
  { option: "--import", capability: "import" },
  { option: "--quit", capability: "quit" },
  { option: "--quit-after", capability: "quitAfter" },
  { option: "--lsp-port", capability: "lsp" },
  { option: "--dap-port", capability: "dap" },
  { option: "--debug-server", capability: "debugServer" },
  { option: "--build-solutions", capability: "buildSolutions" },
  { option: "--dump-extension-api", capability: "extensionApiDump" },
  { option: "--dump-extension-api-with-docs", capability: "extensionApiWithDocsDump" },
  { option: "--validate-extension-api", capability: "extensionApiValidation" },
  { option: "--doctool", capability: "docTool" },
  { option: "--write-movie", capability: "movieWriting" },
];

/**
 * Option tokens that must never be passed to a Godot probe executable.
 * Fixed Solaris probes pass only `--version`, `--help`, or
 * `--dump-extension-api`; these project-affecting tokens are prohibited in
 * probe invocation code and used by the architecture guardrail.
 */
export const FORBIDDEN_GODOT_PROJECT_ARGUMENTS: readonly string[] = [
  "--path",
  "--upwards",
  "--import",
  "--scene",
  "--script",
];

File: crates/siralos-core/src/godot/capabilities.rs
//! Command capabilities advertised by a Godot executable's `--help` output.
//!
//! Presence means advertised support, not operationally verified support.
//!
//! The two states stay distinct (see `verified_capabilities` on the engine
//! profile).

/// Capabilities advertised via `--help`.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct GodotCommandCapabilities {
    /// `--editor`.
    pub editor: bool,
    /// `--project-manager`.
    pub project_manager: bool,
    /// `--recovery-mode`.
    pub recovery_mode: bool,
    /// `--headless`.
    pub headless: bool,
    /// `--path`.
    pub project_path: bool,
    /// `--scene`.
    pub scene: bool,
    /// `--script`.
    pub script: bool,
    /// `--check-only`.
    pub check_only: bool,
    /// `--import`.
    pub import: bool,
    /// `--quit`.
    pub quit: bool,
    /// `--quit-after`.
    pub quit_after: bool,
    /// `--lsp-port`.
    pub lsp: bool,
    /// `--dap-port`.
    pub dap: bool,
    /// `--debug-server`.
    pub debug_server: bool,
    /// `--build-solutions`.
    pub build_solutions: bool,
    /// `--dump-extension-api`.
    pub extension_api_dump: bool,
    /// `--dump-extension-api-with-docs`.
    pub extension_api_with_docs_dump: bool,
    /// `--validate-extension-api`.
    pub extension_api_validation: bool,
    /// `--doctool`.
    pub doc_tool: bool,
    /// `--write-movie`.
    pub movie_writing: bool,
}

/// Create an empty capability set with all flags cleared.
pub fn empty_godot_command_capabilities() -> GodotCommandCapabilities {
    GodotCommandCapabilities {
        editor: false,
        project_manager: false,
        recovery_mode: false,
        headless: false,
        project_path: false,
        scene: false,
        script: false,
        check_only: false,
        import: false,
        quit: false,
        quit_after: false,
        lsp: false,
        dap: false,
        debug_server: false,
        build_solutions: false,
        extension_api_dump: false,
        extension_api_with_docs_dump: false,
        extension_api_validation: false,
        doc_tool: false,
        movie_writing: false,
    }
}

#[cfg(test)]
mod tests {
    use super::empty_godot_command_capabilities;

    #[test]
    fn empty_capabilities_are_all_false() {
        let caps = empty_godot_command_capabilities();
        assert!(!caps.editor);
        assert!(!caps.extension_api_dump);
        assert!(!caps.movie_writing);
    }
}
